modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java

```diff
@@ -552,4 +554,138 @@ public static String decodeBase64(String receiver) {
+            String format = "Extra '.' in path [%s] at index [%d]";
+            throw new IllegalArgumentException(String.format(Locale.ROOT, format, String.join(".", elements), i));
+        } else if (obj instanceof Map<?,?>) {
+            return getByPath((Map<?,?>) obj, elements, i, defaultSupplier);
+        } else if (obj instanceof List<?>) {
+            return getByPath((List<?>) obj, elements, i, defaultSupplier);
+        }
+        return handleMissing(obj, elements, i, defaultSupplier);
+    }
+
+    /**
+     * Map version of getByPath, calls dispatch function on next path element with object map[elements[i]] or
+     * return defaultSupplier if element is missing an i is the last index of elements.
+     * Throws 'IllegalArgumentException' is element is missing and i is not the last index of elements.
+     */
+    private static <K,V> Object getByPath(Map<K,V> map, String[] elements, int i, Supplier<Object> defaultSupplier) {
+        String element = elements[i];
+        if (map.containsKey(element)) {
+            return getByPath(map.get(element), elements, i + 1, defaultSupplier);
+        }
+        return handleMissing(map, elements, i, defaultSupplier);
+    }
+
+
+    /**
+     * List version of getByPath, calls dispatch function on next path element with object list[elements[i]] or
+     * return defaultSupplier if element is missing an i is the last index of elements.
+     * Throws 'IllegalArgumentException' is element is missing and i is not the last index of elements.
+     * Throws 'IllegalArgumentException' with `NumberFormatException' as cause if elements[i] cannot be parsed
+     * as an int.
+     */
+    private static <E> Object getByPath(List<E> list, String[] elements, int i, Supplier<Object> defaultSupplier) {
+        String element = elements[i];
+        try {
+            int elemInt = Integer.parseInt(element);
+            if (list.size() >= elemInt) {
+                return getByPath(list.get(elemInt), elements, i + 1, defaultSupplier);
+            }
+        } catch (NumberFormatException e) {
+            String format = "Could not parse [%s] as a int index into list at path [%s] and index [%d]";
+            throw new IllegalArgumentException(String.format(Locale.ROOT, format, element, String.join(".", elements), i), e);
+        }
+        return handleMissing(list, elements, i, defaultSupplier);
+    }
+
+    /**
+     * Splits path by '.'
+     * Throws 'IllegalArgumentException' if no elements are in the path or if path ends with '.'
+     */
+    private static String[] splitPath(String path) {
+        if (path.endsWith(".")) {
+            String format = "Trailing '.' in path [%s]";
+            throw new IllegalArgumentException(String.format(Locale.ROOT, format, path));
+        }
+        String[] elements = path.split("\\.");
+        if (elements.length == 1 && elements[0].length() == 0) {
+            throw new IllegalArgumentException("Missing path");
```

Comment:
We could check for this before actually splitting right? It would just be empty path?

Reply:
Yup, moved that check up.